tutorials/scripting/c_sharp/c_sharp_signals.rst

```diff
@@ -0,0 +1,155 @@
+``SignalName`` class. This is useful when, for example, you want to await on a signal (see :ref:`doc_c_sharp_differences_yield`).
+
+.. code-block:: csharp
+
+    await ToSignal(GetTree(), SceneTree.SignalName.ProcessFrame);
+
+.. note::
+
+    Godot will take care of disconnecting all the signals you connected through events when your
+    nodes are freed. Meaning that: as you don't need to call ``Disconnect`` on all signals you used
+    ``Connect`` on, you don't need to ``-=`` all the signals you used ``+=`` on.
+
+Custom signals as C# events
+---------------------------
+
+To declare a custom event in your C# script, use the ``[Signal]`` attribute on a delegate type.
+Note that the name of this delegate needs to end with ``EventHandler``.
+
+.. code-block:: csharp
+
+    [Signal]
+    public delegate void MySignalEventHandler();
+
+    [Signal]
+    public delegate void MySignalWithArgumentEventHandler(string myString);
+
+Once this is done, Godot will create the appropriate events automatically behind the scenes. You
+can then use said events as you'd do for any other Godot signal. Note that events are named using
+your delegate's name minus the final ``EventHandler`` part.
+
+.. code-block:: csharp
+
+    public override void _Ready()
+    {
+        MySignal += () => GD.Print("Hello!");
+        MySignalWithArgument += SayHelloTo;
+    }
+
+    private void SayHelloTo(string name)
+    {
+        GD.Print($"Hello {name}!");
+    }
+
+.. warning::
+
+    If you want to connect to these signals in the editor, you will need to (re)build the project
+    to see them appear.
+
+    You can click the **Build** button in the upper-right corner of the editor to do so.
+
+Signal emission
+---------------
+
+To emit signals, use the ``EmitSignal`` method. Note that, as for signals defined by the engine,
+your custom signal names are listed under the nested ``SignalName`` class.
+
+.. code-block:: csharp
+
+    public void MyMethodEmittingSignals()
+    {
+        EmitSignal(SignalName.MySignal);
+        EmitSignal(SignalName.MySignalWithArgument, "World");
+    }
+
+Signals support arguments of:
+
+* All the `built-in value types <https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/built-in-types-table>`_,
+  except ``decimal``, ``nint`` and ``nuint``
+* ``string``
+* Classes derived from :ref:`Godot.Object <class_Object>`
+* Collections types defined in the ``Godot.Collections`` namespace
+
+Consequently, any ``Node`` or ``Reference`` will be compatible automatically, but custom data objects will need
+to inherit from ``Godot.Object`` or one of its subclasses.
+
+.. code-block:: csharp
+
+    public partial class DataObject : Godot.Object
+    {
+        public string MyFirstString { get; set; }
+        public string MySecondString { get; set; }
+    }
+
+.. warning::
+
+    You should not use ``Invoke`` on events tied to Godot signals, as this would not execute the
```

Comment:
Just to be clear, you can't use `Invoke` even if you wanted to, because of the way we generate the events.